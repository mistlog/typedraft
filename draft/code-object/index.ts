export * from "./export-class";
export * from "./inline-context";
export * from "./local-context";
export * from "./method";
export * from "./module";
