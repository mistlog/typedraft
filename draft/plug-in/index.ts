export * from "./draft-plugin-class";
export * from "./draft-plugin-dsl";
export * from "./draft-plugin-filter";
export * from "./draft-plugin-inplace-context";
export * from "./draft-plugin-local-context";
export * from "./draft-plugin-refresh";
