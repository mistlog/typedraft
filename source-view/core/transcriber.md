# Transcriber

We always write code in one language and translate(compile) it to another, however, in both language, the code we write directly represent the execution flow.

When we write draft, we don't expect it executes as is, instead, it's designed to be transformed, not execute. Code is only meaningful within its' context. That's why we choose another word to differentiate it from compiler.

```typescript
export class Transcriber {
    //
    m_Module: ModuleCode;

    // the 3 types of code we will transform
    m_ClassMap: Map<string, ExportClassCode>;
    m_MethodMap: Map<string, Array<ClassMethod>>;
    m_ContextMap: Map<string, LocalContext>;

    //
    m_DSLMap: Map<string, IDSL>;
    m_Plugins: Array<IPlugin>;

    //
    get m_Code() {
        return this.m_Module.m_Code;
    }
    set m_Code(code: Array<Statement>) {
        this.m_Module.m_Code = code;
    }
}
```

A transcriber just likes a container, it's a collection of plugins. For example, to transform local context, we have ```draft-plugin-local-context```, to add class method to class, we have ```draft-plugin-class```, and ```draft-plugin-dsl``` for DSL extension support.

```typescript
<Transcriber /> +
    function Transcribe(this: Transcriber) {
        this.m_Plugins.forEach(plugin => plugin.Transcribe());
        return ToString(this.m_Module.m_File, { comments: false });
    };
```

# Preprocess

```typescript
export interface ITranscriber {
    Preprocess: () => void;
}
```

When we init a transcriber, we will preprocess code to build some maps for lookup purpose.

```typescript
<Transcriber /> +
    function Preprocess(this: Transcriber) {
        const draft = this.m_Module.ToDraft();
        draft.forEach(each => {
            if (each instanceof ExportClassCode) {
                this.m_ClassMap.set(each.m_Name, each);
            } else if (each instanceof MethodCode) {
                const { class_name, method } = each.ToClassMethod();
                const methods = this.m_MethodMap.get(class_name);
                if (!methods) {
                    this.m_MethodMap.set(class_name, [method]);
                } else {
                    methods.push(method);
                }
            } else if (each instanceof LocalContext) {
                this.m_ContextMap.set(each.m_Name, each);
            }
        });
    };
```

# Utility

Other methods are just utility.

## DSL

```typescript
export interface IDSL {
    Transcribe(block: Array<Statement>): Array<Statement>;
}
```

```typescript
export interface ITranscriber {
    GetDSL: (name: string) => IDSL;
    PrepareDSLs: () => void;
}
```

## Plugin

```typescript
export interface IPlugin {
    Transcribe(): void;
}
```

```typescript
export interface ITranscriber {
    PreparePlugins: () => void;
}
```

## Local Context

```typescript
export interface ITranscriber {
    TraverseLocalContext: (callback: ITraverseLocalContextCallback) => void;
    GetLocalContext: (name: string) => LocalContext;
}
```

```typescript
export type ITraverseLocalContextCallback = (context: LocalContext, name: string) => void;
```

## Class

```typescript
export interface ITranscriber {
    GetClass: (name: string) => ExportClassCode;
    TraverseMethod: (callback: ITraverseMethodCallback) => void;
}
```

```typescript
export type ITraverseMethodCallback = (methods: Array<ClassMethod>, class_name: string) => void;
```

# Implementation

##  DSL

```typescript
<Transcriber /> +
    function AddDSL(this: Transcriber, name: string, dsl: IDSL) {
        this.m_DSLMap.set(name, dsl);
    };
```

```typescript
<Transcriber /> +
    function GetDSL(this: Transcriber, name: string) {
        return this.m_DSLMap.get(name);
    };
```

```typescript
<Transcriber /> +
    function PrepareDSLs(this: Transcriber) {
        this.m_DSLMap.set("match", new PatternMatch());
    };
```

##  Plugin

```typescript
<Transcriber /> +
    function PreparePlugins(this: Transcriber & ITranscriber) {
        /**
         * pay attention to the default order of plugins
         * 1. resolve DSL first
         * 2. then context
         * 3. add methods to class
         * 4. remove redundant code
         */
        this.m_Plugins = [new DSLPlugin(this), new LocalContextPlugin(this), new ClassPlugin(this), new FilterPlugin(this)];
    };
```

```typescript
<Transcriber /> +
    function AddPlugin(plugin: IPlugin) {
        this.m_Plugins.push(plugin);
    };
```

##  Local Context

```typescript
<Transcriber /> +
    function GetLocalContext(this: Transcriber, name: string) {
        return this.m_ContextMap.get(name);
    };
```

```typescript
<Transcriber /> +
    function TraverseLocalContext(this: Transcriber, callback: ITraverseLocalContextCallback) {
        this.m_ContextMap.forEach((context, name) => callback(context, name));
    };
```

##  Class

```typescript
<Transcriber /> +
    function GetClass(this: Transcriber, name: string) {
        return this.m_ClassMap.get(name);
    };
```

```typescript
<Transcriber /> +
    function TraverseMethod(this: Transcriber, callback: ITraverseMethodCallback) {
        this.m_MethodMap.forEach((methods, class_name) => callback(methods, class_name));
    };
```

# Trivial

```typescript
export interface ITranscriber {
    m_Code: Array<Statement>;
}
```

```typescript
<Transcriber /> +
    function constructor(this: Transcriber & ITranscriber, _module: ModuleCode | string) {
        //
        this.m_Module = typeof _module === "string" ? new ModuleCode(_module) : _module;

        //
        this.m_ClassMap = new Map();
        this.m_MethodMap = new Map();
        this.m_ContextMap = new Map();
        this.m_DSLMap = new Map();

        //
        this.PrepareDSLs();
        this.PreparePlugins();

        //
        this.Preprocess();
    };
```