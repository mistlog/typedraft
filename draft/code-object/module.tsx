import {
    Program,
    File,
    ExpressionStatement,
    ExportNamedDeclaration,
    FunctionDeclaration,
} from "@babel/types";
import { ExportClassCode } from "./export-class";
import { MethodCode } from "./method";
import { LocalContext } from "./local-context";
import { ToFile, ToString } from "../common/utility";
import traverse, { NodePath, Node, TraverseOptions } from "@babel/traverse";
import { InlineContext } from "./inline-context";
import { MatchDSL, __ } from "draft-dsl-match";

/**
 * # Draft
 * A .tsx file is considered as a module and we will transform these 4 types of code, they are collected and then transcibed to "real" code.
 */
export class ModuleCode {
    m_File: File;
    m_Path: NodePath<Program>;

    constructor(code: string) {
        this.m_File = ToFile(code);
    }

    /**
     * As we are only interested in the draft part of a module, then we need a way to return this "view" of module code.
     */
    ToDraft() {
        /**
         * refresh and update bindings because DSL only transforms code
         */
        this.m_File = ToFile(ToString(this.m_File));

        /**
         * traverse file and set path
         */
        let draft: Draft = [];
        this.Traverse<{ _module: ModuleCode }>(
            {
                Program(path) {
                    this._module.m_Path = path;
                },

                ExpressionStatement(path) {
                    const expression = path.get("expression");
                    if (expression.isStringLiteral()) {
                        const literal = expression.node.value.trim();
                        if (literal.startsWith("use") && path.parentPath.isBlockStatement()) {
                            /**
                             * inline context should be resolved from deepest
                             */
                            draft.unshift(new InlineContext(path.parentPath));
                        }
                    }
                },
            },
            { _module: this }
        );

        /**
         * collect draft parts
         */
        this.m_Path.get("body").forEach(path => {
            /**
             * remove redundant ; before tag, see comment of `ToFile` in utility.ts
             */
            if (path.isEmptyStatement()) {
                path.remove();
                return;
            }

            // prettier-ignore
            const to_add = Î›<ExportClassCode | LocalContext | MethodCode | null>("match")` ${path as NodePath<any>} 
                ${IsExportClassCode(path)} -> ${(path: NodePath<ExportNamedDeclaration>) => new ExportClassCode(path)}
                ${IsMethodCode(path)} -> ${(path: NodePath<ExpressionStatement>) => new MethodCode(path)}
                ${IsLocalContext(path)} -> ${(path: NodePath<FunctionDeclaration>) => new LocalContext(path.scope.parent.getBinding(path.node.id.name))}
                ${__} -> ${null}
            `;

            if (to_add) {
                draft.push(to_add);
            }
        });
        return draft;
    }

    Traverse<S>(this: ModuleCode, visitor: TraverseOptions<S>, state: S = null) {
        this.m_File = ToFile(ToString(this.m_File));
        traverse<S>(this.m_File, visitor, null, state);
    }
}

export type Draft = Array<ExportClassCode | MethodCode | LocalContext | InlineContext>;

export function IsExportClassCode(path: NodePath<any>) {
    return (
        path.isExportNamedDeclaration() &&
        (path.get("declaration") as NodePath<Node>).isClassDeclaration()
    );
}

export function IsMethodCode(path: NodePath<any>) {
    if (!path.isExpressionStatement()) {
        return false;
    }

    const expression = path.get("expression") as NodePath<Node>;
    if (!expression.isBinaryExpression()) {
        return false;
    }

    const left = expression.get("left") as NodePath<Node>;
    const right = expression.get("right") as NodePath<Node>;
    return left.isJSXElement() && right.isFunctionExpression();
}

export function IsLocalContext(path: NodePath<any>) {
    if (!path.isFunctionDeclaration()) {
        return false;
    }

    const [directive]: [string] = path.node.body.directives;

    const is_local_context = path.scope.parent
        .getBinding(path.node.id.name)
        .referencePaths.some(path => {
            const used_as_jsx = path.parentPath?.parentPath?.isJSXElement();
            const used_as_statement = path.parentPath?.parentPath?.parentPath?.isExpressionStatement();
            const has_context = Boolean(directive);
            return used_as_jsx && (has_context || used_as_statement);
        });
    return is_local_context;
}
