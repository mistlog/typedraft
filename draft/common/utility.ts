import { template } from "@babel/core";
import { Node } from "@babel/types";
import { NodePath } from "@babel/traverse";
import { default as Binding } from "@babel/traverse/lib/scope/binding";
import { parse } from "@babel/parser";
import generate, { GeneratorOptions } from "@babel/generator";

export function ToAst<T = any>(code: string) {
    return (template.ast(code, { plugins: ["typescript", "jsx"] }) as unknown) as T;
}

export function ToNodePath<T = any>(code: string): NodePath<T> {
    const path = new NodePath<T>(null, null);
    path.node = (ToAst(code) as unknown) as T;
    return path;
}

/**
 * only used in test to construct required binding param
 */
export function ToBinding(code: string): Binding {
    const binding = new Binding({
        path: ToNodePath<Node>(code),
        existing: null,
        identifier: null,
        scope: null,
        kind: "let",
    });
    return binding;
}

export function ToFile(raw: string) {
    // babel bug: https://github.com/babel/babel/issues/8837
    // refert to test case "transcriber.interface-no-parse-error"
    const code = raw.replace(new RegExp("^<", "gm"), ";<");
    return parse(code, {
        sourceType: "module",
        plugins: ["typescript", "jsx", ["decorators", { decoratorsBeforeExport: true }]],
    });
}

export function ToString(node: Node, options?: GeneratorOptions) {
    return generate(node, options || {}).code;
}
