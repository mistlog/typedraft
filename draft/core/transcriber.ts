import {
    Statement,
    FunctionDeclaration,
    BlockStatement,
    Expression,
    TemplateLiteral,
} from "@babel/types";
import { NodePath } from "@babel/traverse";
import { ToString } from "../common/utility";
import { ModuleCode } from "../code-object/module";
import { ExportClassCode } from "../code-object/export-class";
import { MethodCode } from "../code-object/method";
import { LocalContext } from "../code-object/local-context";
import { InlineContext } from "../code-object/inline-context";

/**
 * # Transcriber
 * We always write code in one language and translate(compile) it to another, however, in both language, the code we write directly represents the execution flow.
 * When we write draft, we don't expect it executes as is, instead, it's designed to be transformed, not executed.
 * That's why we choose another word to differentiate it from compiler.
 */
export class Transcriber {
    m_Module: ModuleCode;

    /**
     * the 4 types of code we will transform
     */
    m_ClassMap: Map<string, ExportClassCode>;
    m_MethodMap: Map<string, Array<MethodCode>>;
    m_ContextMap: Map<string, LocalContext>;
    m_InlineContextMap: Map<Symbol, InlineContext>;

    m_DSLMap: Map<string, IDSL>;
    m_Plugins: Array<IPlugin>;

    constructor(_module: string) {
        this.m_Module = new ModuleCode(_module);

        this.m_ClassMap = new Map<string, ExportClassCode>();
        this.m_MethodMap = new Map<string, Array<MethodCode>>();
        this.m_ContextMap = new Map<string, LocalContext>();
        this.m_InlineContextMap = new Map<Symbol, InlineContext>();
        this.m_DSLMap = new Map<string, IDSL>();
        this.m_Plugins = new Array<IPlugin>();
    }

    /**
     * A transcriber is just a container, it's a collection of plugins.
     * For example, to transform local context, we have `draft-plugin-local-context`, to add class method to class, we have `draft-plugin-class`, and `draft-plugin-dsl` for DSL extension support.
     */

    Transcribe() {
        this.m_Plugins.forEach(plugin => plugin.Transcribe());
        return ToString(this.m_Module.m_File, {
            comments: false,
        });
    }

    AddDSL(name: string, dsl: IDSL) {
        this.m_DSLMap.set(name, dsl);
    }

    GetDSL(name: string) {
        return this.m_DSLMap.get(name);
    }

    GetLocalContext(name: string) {
        return this.m_ContextMap.get(name);
    }

    TraverseLocalContext(callback: TraverseLocalContextCallback) {
        this.m_ContextMap.forEach((context, name) => callback(context, name));
    }

    TraverseInlineContext(callback: TraverseInlineContextCallback) {
        this.m_InlineContextMap.forEach(context => callback(context));
    }

    GetClass(name: string) {
        return this.m_ClassMap.get(name);
    }

    TraverseMethod(callback: TraverseMethodCallback) {
        this.m_MethodMap.forEach((methods, class_name) => callback(methods, class_name));
    }
}

export interface IDSL {
    Transcribe?(
        block: Array<Statement>,
        path?: NodePath<FunctionDeclaration> | NodePath<BlockStatement>
    ): Array<Statement>;

    InplaceTranscribe?(literal: NodePath<TemplateLiteral>, type: string): Expression;

    m_Merge?: boolean;
}

export interface IPlugin {
    Transcribe(): void;
}

export interface ITranscriber {
    Transcribe: () => string;

    AddDSL: (name: string, dsl: IDSL) => void;

    GetDSL: (name: string) => IDSL;
    GetClass: (name: string) => ExportClassCode;
    GetLocalContext: (name: string) => LocalContext;

    TraverseMethod: (callback: TraverseMethodCallback) => void;
    TraverseLocalContext: (callback: TraverseLocalContextCallback) => void;
    TraverseInlineContext: (callback: TraverseInlineContextCallback) => void;

    m_Module: ModuleCode;

    /**
     * the 3 types of code we will transform
     */
    m_ClassMap: Map<string, ExportClassCode>;
    m_MethodMap: Map<string, Array<MethodCode>>;
    m_ContextMap: Map<string, LocalContext>;
    m_InlineContextMap: Map<Symbol, InlineContext>;

    m_DSLMap: Map<string, IDSL>;
    m_Plugins: Array<IPlugin>;
}

export type TraverseLocalContextCallback = (context: LocalContext, name: string) => void;
export type TraverseInlineContextCallback = (context: InlineContext) => void;
export type TraverseMethodCallback = (methods: Array<MethodCode>, class_name: string) => void;
