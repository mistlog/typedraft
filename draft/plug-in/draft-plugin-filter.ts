import { ITranscriber } from "../core";

export class FilterPlugin {
    m_Transcriber: ITranscriber;

    constructor(transcriber: ITranscriber) {
        this.m_Transcriber = transcriber;
    }

    Transcribe() {
        this.m_Transcriber.TraverseLocalContext(context => context.m_Path.remove());
        this.m_Transcriber.TraverseMethod(methods => methods.forEach(each => each.m_Path.remove()));
    }
}
