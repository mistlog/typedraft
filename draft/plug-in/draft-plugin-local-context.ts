import { ITranscriber, IPlugin } from "../core";
import { FunctionDeclaration, FunctionExpression } from "@babel/types";
import { NodePath } from "@babel/core";
import * as toposort from "toposort";

export class LocalContextPlugin {
    m_Transcriber: ITranscriber;

    Transcribe() {
        const to_transcribe = this.GetContextList();
        to_transcribe.forEach(name => {
            const context = this.m_Transcriber.GetLocalContext(name);
            context.m_Refs.forEach(path =>
                path
                    .findParent(path => path.isExpressionStatement())
                    .replaceWithMultiple(context.ToStatements())
            );
        });
    }

    GetContextList() {
        const graph = [];
        this.m_Transcriber.TraverseLocalContext((context, name) => {
            context.m_Refs.forEach(path => {
                const parent = path.findParent(
                    path =>
                        path.isFunctionDeclaration() ||
                        path.isFunctionExpression() ||
                        path.isExportDeclaration()
                ) as NodePath<FunctionDeclaration | FunctionExpression>;

                if (parent) {
                    graph.push([name, parent.node.id.name]);
                }
            });
        });
        const context_list = toposort(graph).filter(name =>
            this.m_Transcriber.GetLocalContext(name)
        );
        return context_list;
    }

    constructor(transcriber: ITranscriber) {
        this.m_Transcriber = transcriber;
    }
}

export class ILocalContextPlugin implements IPlugin {
    Transcribe: () => void;
    GetContextList: () => Array<string>;
}
