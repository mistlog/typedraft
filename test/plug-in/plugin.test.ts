import { ModuleCode } from "../../src/code-object/module";
import { Transcriber, IDSL } from "../../src/core/transcriber";
import { LocalContextPlugin } from "../../src/plug-in/draft-plugin-local-context";
import { ToAst, ToString } from "../../src/common/utility";
import { FilterPlugin } from "../../src/plug-in/draft-plugin-filter";
import { Statement, FunctionDeclaration } from "@babel/types";
import { NodePath } from "@babel/traverse";

describe("plugin.local-context", () =>
{
    test("transcribe.simple", () =>
    {
        //
        const code = `
            <Foo/> + function Test(this: Foo, a: number, b: string){
                <Snippet/>;
                return a.toString()+b;
            };

            function Snippet(){
                console.log("hi");
            }
        `;

        const transcriber = new Transcriber(code);

        // remove last filter plugin from default transcriber
        transcriber.m_Plugins.pop();
        const plugin = transcriber.m_Plugins.find(plugin => plugin instanceof LocalContextPlugin);
        Reflect.set(plugin, "GetContextList", () => ["Snippet"]);

        const result = transcriber.Transcribe();
        expect(result).toMatchSnapshot();
    })
    test("transcribe.nested", () =>
    {
        //
        const code = `
            <Foo/> + function Test(this: Foo, a: number, b: string){
                <Snippet/>;
                return a.toString()+b;
            };

            function Snippet(){
                console.log("hi");
                <SnippetNested/>;
            }

            function SnippetNested(){
                console.log("nested");
            }
        `;

        const transcriber = new Transcriber(code);
        transcriber.m_Plugins.pop();
        const plugin = transcriber.m_Plugins.find(plugin => plugin instanceof LocalContextPlugin);
        Reflect.set(plugin, "GetContextList", () => ["SnippetNested", "Snippet"]);

        const result = transcriber.Transcribe();;
        expect(result).toMatchSnapshot();
    })


    test("transcribe.integrated.export-function", () =>
    {
        //
        const code = `
            export function Test(a: number, b: string){
                <Snippet/>;
                return a.toString()+b;
            };

            function Snippet(){
                console.log("export function");
            }
        `;

        const transcriber = new Transcriber(code);
        transcriber.m_Plugins.pop();
        const result = transcriber.Transcribe();
        expect(result).toMatchSnapshot();
    })

    test("integrated.method", () =>
    {
        //
        const code = `
            <Foo/> + function Test(this: Foo, a: number, b: string){
                <Snippet/>;
                <AnotherSnippet/>;
                return a.toString()+b;
            };

            function Snippet(){
                console.log("hi");
                <SnippetNested/>;
            }

            function AnotherSnippet(){
                console.log("another");
                <SnippetNested/>;
            }

            function SnippetNested(){
                console.log("nested");
                <DeepSnippet/>;
            }

            function DeepSnippet(){
                console.log("deep");
            }
        `;

        const transcriber = new Transcriber(code);
        transcriber.m_Plugins.pop();
        const result = transcriber.Transcribe();
        expect(result).toMatchSnapshot();
    })

    test("get-context-list.simple", () =>
    {
        //
        const code = `
            <Foo/> + function Test(this: Foo, a: number, b: string){
                <Snippet/>;
                return a.toString()+b;
            };

            function Snippet(){
                console.log("hi");
            }
        `;

        const transcriber = new Transcriber(code);
        const plugin = transcriber.m_Plugins.find(plugin => plugin instanceof LocalContextPlugin) as LocalContextPlugin;
        const context_list = plugin.GetContextList();
        expect(context_list).toEqual(["Snippet"]);
    })

    test("get-context-list.nested", () =>
    {
        //
        const code = `
            <Foo/> + function Test(this: Foo, a: number, b: string){
                <Snippet/>;
                <AnotherSnippet/>;
                return a.toString()+b;
            };

            function Snippet(){
                console.log("hi");
                <SnippetNested/>;
            }

            function AnotherSnippet(){
                console.log("another");
                <SnippetNested/>;
            }

            function SnippetNested(){
                console.log("nested");
                <DeepSnippet/>;
            }

            function DeepSnippet(){
                console.log("deep");
            }
        `;

        const transcriber = new Transcriber(code);
        transcriber.m_Plugins.pop();
        const plugin = transcriber.m_Plugins.find(plugin => plugin instanceof LocalContextPlugin) as LocalContextPlugin;
        const context_list = plugin.GetContextList();
        expect(context_list).toEqual(["DeepSnippet", "SnippetNested", "Snippet", "AnotherSnippet"]);
    })
})

describe("plugin.dsl", () =>
{
    test("dsl.log", () =>
    {
        //
        class Foo implements IDSL
        {
            Transcribe(block: Array<Statement>): Array<Statement>
            {

                const transcribed = ToAst(`
                    console.log("current");
                `);

                return [transcribed as Statement];
            }
        }

        //
        const code = `
            function Test(){
                'use foo';
                console.log("previous");
            }
        `;

        const transcriber = new Transcriber(code);
        transcriber.m_Plugins.pop();
        transcriber.AddDSL("foo", new Foo());
        const result = transcriber.Transcribe();

        //
        const expected = ToAst(`
            function Test(){
                console.log("current");
            }
        `);

        expect(ToAst(result)).toEqual(expected);
    })

    test("dsl.rename.test-use-path", () =>
    {
        //
        class Foo implements IDSL
        {
            Transcribe(block: Array<Statement>, path: NodePath<FunctionDeclaration>): Array<Statement>
            {

                path.traverse({
                    Identifier(path)
                    {
                        if (path.node.name === "x")
                        {
                            path.node.name = "y";
                        }
                    }
                })

                return block;
            }
        }

        //
        const code = `
            function Test(){
                'use foo';
                const x = 1;
                console.log(x + 1);
            }
        `;

        const transcriber = new Transcriber(code);
        transcriber.m_Plugins.pop();
        transcriber.AddDSL("foo", new Foo());
        const result = transcriber.Transcribe();

        //
        const expected = ToAst(`
            function Test(){
                const y = 1;
                console.log(y + 1);
            }
        `);

        expect(ToAst(result)).toEqual(expected);
    })
})

describe("plugin.filter", () =>
{
    test("filter", () =>
    {
        //
        const code = `
            export interface IFoo{
                foo: number;
            }

            export class Foo{

            }

            <Foo/> + function foo(){
                <Snippet/>;
            };

            function Snippet(){
                console.log("hi");
            }
        `;

        const transcriber = new Transcriber(code);
        transcriber.m_Plugins.pop();
        transcriber.AddPlugin(new FilterPlugin(transcriber));
        const result = transcriber.Transcribe();
        expect(result).toMatchSnapshot();
    })
})