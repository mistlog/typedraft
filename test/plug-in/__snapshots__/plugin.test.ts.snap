// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin.filter filter 1`] = `
"export interface IFoo {
  foo: number;
}
export class Foo {
  foo() {
    ;
    console.log(\\"hi\\");
  }

}
;"
`;

exports[`plugin.local-context integrated.method 1`] = `
";
<Foo /> + function Test(this: Foo, a: number, b: string) {
  ;
  console.log(\\"hi\\");
  console.log(\\"nested\\");
  console.log(\\"deep\\");
  console.log(\\"another\\");
  console.log(\\"nested\\");
  console.log(\\"deep\\");
  return a.toString() + b;
};

function Snippet() {
  console.log(\\"hi\\");
  console.log(\\"nested\\");
  console.log(\\"deep\\");
}

function AnotherSnippet() {
  console.log(\\"another\\");
  console.log(\\"nested\\");
  console.log(\\"deep\\");
}

function SnippetNested() {
  console.log(\\"nested\\");
  console.log(\\"deep\\");
}

function DeepSnippet() {
  console.log(\\"deep\\");
}"
`;

exports[`plugin.local-context transcribe.integrated.export-function 1`] = `
"export function Test(a: number, b: string) {
  ;
  console.log(\\"export function\\");
  return a.toString() + b;
}

function Snippet() {
  console.log(\\"export function\\");
}"
`;

exports[`plugin.local-context transcribe.nested 1`] = `
";
<Foo /> + function Test(this: Foo, a: number, b: string) {
  ;
  console.log(\\"hi\\");
  console.log(\\"nested\\");
  return a.toString() + b;
};

function Snippet() {
  console.log(\\"hi\\");
  console.log(\\"nested\\");
}

function SnippetNested() {
  console.log(\\"nested\\");
}"
`;

exports[`plugin.local-context transcribe.simple 1`] = `
";
<Foo /> + function Test(this: Foo, a: number, b: string) {
  ;
  console.log(\\"hi\\");
  return a.toString() + b;
};

function Snippet() {
  console.log(\\"hi\\");
}"
`;
